File: app/src/main/java/com/xanarry/lantrans/network/UdpServer.java
package com.xanarry.lantrans.network;

import android.util.Log;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.util.concurrent.TimeoutException;

import static com.xanarry.lantrans.utils.Configuration.BROADCAST_DATA;
import static com.xanarry.lantrans.utils.Configuration.RESPONSE_DATA;

/**
 * Created by xanarry on 2016/5/22.
 */
public class UdpServer {
    public static final String TAG = UdpServer.class.getName();;

    private int mPort;
    private DatagramSocket socket;
    private boolean mLive = true;

    public UdpServer(int port) {
        this.mPort = port;
    }

    public void waitClient() throws IOException {
            socket = new DatagramSocket(mPort);//设置服务器端口, 监听广播信息

            byte[] buf = new byte[BROADCAST_DATA.getBytes().length];
            DatagramPacket message = new DatagramPacket(buf, buf.length);
            while (mLive) {
                message.setData(buf);
                try {
                    socket.receive(message);
                } catch (SocketTimeoutException e) {

                }

                String msg = new String(message.getData());

                Log.e(TAG, "waitClient: " + msg);
                if (BROADCAST_DATA.equals(msg)) {
                    message.setData(RESPONSE_DATA.getBytes());
                    socket.send(message);
                    Log.e(TAG, "waitClient: send out");
                }
            }
        socket.close();
    }

    public void kill() {
        mLive = false;
    }

    public void close() {
        socket.close();
    }
}

File: app/src/main/java/com/xanarry/lantrans/network/UdpServer.kt
package com.xanarry.lantrans.network

import android.util.Log
import com.xanarry.lantrans.utils.Configuration.BROADCAST_DATA
import com.xanarry.lantrans.utils.Configuration.RESPONSE_DATA
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.SocketTimeoutException

/**
 *
 */
class UdpServer(private val mPort: Int) {
    private var socket: DatagramSocket? = null
    private var mLive = true

    @Throws(IOException::class)
    fun waitClient() {
        socket = DatagramSocket(mPort)

        val buf = ByteArray(BROADCAST_DATA.toByteArray().size)
        val message = DatagramPacket(buf, buf.size)
        while (mLive) {
            message.data = buf
            try {
                socket!!.receive(message)
            } catch (e: SocketTimeoutException) {

            }

            val msg = String(message.data)

            Log.e(TAG, "waitClient: $msg")
            if (BROADCAST_DATA == msg) {
                message.data = RESPONSE_DATA.toByteArray()
                socket!!.send(message)
                Log.e(TAG, "waitClient: send out")
            }
        }
        socket!!.close()
    }

    fun kill() {
        mLive = false
    }

    fun close() {
        socket!!.close()
    }

    companion object {
        const val TAG = "UdpServer"
    }
}
